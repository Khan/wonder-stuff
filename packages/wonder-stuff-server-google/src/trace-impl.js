// @flow
import type {Tracer} from "@google-cloud/trace-agent";
import {getGatewayInfo} from "./get-gateway-info.js";
import {getDelta} from "./get-delta.js";
import {getDefaultMetadata} from "./create-logger.js";
import KAError from "./ka-error.js";
import {Errors} from "./errors.js";
import type {Logger, ITraceSession, TraceSessionInfo} from "./types.js";

/**
 * Start tracing an event.
 *
 * This will log the start of a trace and open a trace session, which is
 * returned. Use the returned session to end the trace when the traced event is
 * over. The traced event will be logged and also written to the Google Cloud
 * StackDriver Trace agent.
 *
 * Trace logs include metadata about the trace such as duration and memory
 * usage.
 *
 * @param {Logger} logger A logger to use for documention and timing the
 * traced action.
 * @param {string} action The name of the traced action. Keep it short. This
 * should be the name of an action rather than a specific URL, for example. Use
 * addLabel on the returned session or the session info when ending the session
 * to add additional details about the trace.
 * @param {string} message A message that will be logged. This is not included
 * in the traces.
 * @param {Tracer} [tracer] A Google Cloud trace agent tracer which
 * can be used to record the traced action.
 * @returns {ITraceSession} A trace session that the caller should use to
 * indicate when the session is finished.
 */
export const traceImpl = (
    logger: Logger,
    action: string,
    message: string,
    tracer?: Tracer,
): ITraceSession => {
    if (!action) {
        throw new KAError(
            "Must provide an action for the trace session.",
            Errors.Internal,
        );
    }
    const logMessage = `${action}${message ? `: ${message}` : ""}`;

    /**
     * We are going to use the logger's profiling API (provided by winston).
     * However, we want to mark the start of the trace as it gives us some
     * debug information which can be valuable when investigating operations.
     *
     * Winston only logs when profiling is done and the optional trace agent
     * tracer will only show the span if it is ended.
     *
     * Since this is noise in most situations, we will log this at the lowest
     * level of silly.
     */
    logger.silly(`TRACE ${logMessage}`);

    /**
     * Now we start the profiling timer.
     */
    const profiler = logger.startTimer();
    const beforeMemory = process.memoryUsage();
    const {name: gatewayName} = getGatewayInfo();

    /**
     * Next, if we were given a tracer, we start a trace section for this so
     * trace session so that it will appear in Stackdriver Trace.
     *
     * We annotate the span with the gateway name so that it is clear in the
     * trace which spans were created by this API and which were inserted by
     * other means.
     */
    const span = tracer?.createChildSpan({name: `${gatewayName}.${action}`});

    const profileLabels = {};
    const addLabel = <T>(name: string, value: T): void => {
        /**
         * Track this so we can also include it in our logging info.
         */
        profileLabels[name] = value;

        /**
         * Send this label on to the trace span.
         *
         * We disable this lint rule as the linter does not appear to
         * understand the optional chaining.
         */
        span?.addLabel(name, value);
    };

    /**
     * This is the function that we will return to our caller.
     * It can then be used to end and record the trace session.
     */
    const end = (info?: TraceSessionInfo): void => {
        const afterMemory = process.memoryUsage();

        /**
         * Add some session information to the span as labels.
         *
         * Flow doesn't trust the inexact objects returned by memoryUsage()
         * and so `getDelta` isn't typed to handle that. Have to rethink
         * how to make that work.
         */
        // $FlowFixMe[incompatible-call]
        addLabel("/memory/delta", getDelta(beforeMemory, afterMemory));
        addLabel("/memory/final", afterMemory);

        /**
         * We need to build the metadata that we will be logging.
         * This is a combination of the given info, some custom things we add,
         * and any profile labels that were added.
         */
        const metadata = {
            /**
             * We have to add the default metadata because winston does not
             * include this for profiler.done calls, strangely.
             *
             * And we have to recreate it because we might be in a worker
             * that doesn't have access directly to the root logger that has
             * the default data.
             */
            ...getDefaultMetadata(),
            ...profileLabels,
            ...info,
            message: `TRACED ${logMessage}`,
            level: info?.level || "debug",
        };

        /**
         * Let's mark our profile as done.
         *
         * We include the session info object, but make sure to set the level
         * and message ourselves.
         */
        profiler.done(metadata);

        /**
         * If we started a tracer span, let's end it.
         *
         * We disable this lint rule as the linter does not appear to
         * understand the optional chaining.
         */
        span?.endSpan();
    };

    return {
        get action() {
            return action;
        },
        addLabel,
        end,
    };
};
