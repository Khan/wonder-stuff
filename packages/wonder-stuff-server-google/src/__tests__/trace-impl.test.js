// @flow
import * as GetGatewayInfo from "../get-gateway-info.js";
import {traceImpl} from "../trace-impl.js";

jest.mock("../get-gateway-info.js");

describe("#trace", () => {
    it.each([[""], [null], [undefined]])(
        "should throw if the action is empty/falsy",
        (testPoint) => {
            // Arrange
            const fakeLogger = ({}: any);
            jest.spyOn(GetGatewayInfo, "getGatewayInfo").mockReturnValue({
                name: "GATEWAY_NAME",
                version: "GATEWAY_VERSION",
            });

            // Act
            const underTest = () => traceImpl(fakeLogger, testPoint, "MESSAGE");

            // Assert
            expect(underTest).toThrowErrorMatchingSnapshot();
        },
    );

    it("should log the start of the session", () => {
        // Arrange
        const fakeLogger = ({
            silly: jest.fn(),
            startTimer: jest.fn(),
        }: any);
        jest.spyOn(GetGatewayInfo, "getGatewayInfo").mockReturnValue({
            name: "GATEWAY_NAME",
            version: "GATEWAY_VERSION",
        });

        // Act
        traceImpl(fakeLogger, "ACTION", "MESSAGE");

        // Assert
        expect(fakeLogger.silly).toHaveBeenCalledWith("TRACE ACTION: MESSAGE");
    });

    it("should start a timer", () => {
        // Arrange
        const fakeLogger = ({
            silly: jest.fn(),
            startTimer: jest.fn(),
        }: any);
        jest.spyOn(GetGatewayInfo, "getGatewayInfo").mockReturnValue({
            name: "GATEWAY_NAME",
            version: "GATEWAY_VERSION",
        });

        // Act
        traceImpl(fakeLogger, "ACTION", "MESSAGE");

        // Assert
        expect(fakeLogger.startTimer).toHaveBeenCalled();
    });

    it("should create a tracer span", () => {
        // Arrange
        const fakeLogger: any = {
            silly: jest.fn(),
            startTimer: jest.fn(),
        };
        const fakeTracer: any = {
            createChildSpan: jest.fn(),
        };
        jest.spyOn(GetGatewayInfo, "getGatewayInfo").mockReturnValue({
            name: "GATEWAY_NAME",
            version: "GATEWAY_VERSION",
        });

        // Act
        traceImpl(fakeLogger, "ACTION", "MESSAGE", fakeTracer);

        // Assert
        expect(fakeTracer.createChildSpan).toHaveBeenCalledWith({
            name: "GATEWAY_NAME.ACTION",
        });
    });

    it("should return a trace session", () => {
        // Arrange
        const fakeLogger: any = {
            silly: jest.fn(),
            startTimer: jest.fn(),
        };
        jest.spyOn(GetGatewayInfo, "getGatewayInfo").mockReturnValue({
            name: "GATEWAY_NAME",
            version: "GATEWAY_VERSION",
        });

        // Act
        const result = traceImpl(fakeLogger, "ACTION", "MESSAGE");

        // Assert
        expect(result).toBeDefined();
    });

    describe("trace session", () => {
        describe("#action", () => {
            it("should give action of the trace", () => {
                // Arrange
                const fakeLogger: any = {
                    silly: jest.fn(),
                    startTimer: jest.fn(),
                };
                jest.spyOn(GetGatewayInfo, "getGatewayInfo").mockReturnValue({
                    name: "GATEWAY_NAME",
                    version: "GATEWAY_VERSION",
                });
                const session = traceImpl(fakeLogger, "ACTION", "MESSAGE");

                // Act
                const result = session.action;

                // Assert
                expect(result).toBe("ACTION");
            });

            it("should be read-only", () => {
                // Arrange
                const fakeLogger: any = {
                    silly: jest.fn(),
                    startTimer: jest.fn(),
                };
                jest.spyOn(GetGatewayInfo, "getGatewayInfo").mockReturnValue({
                    name: "GATEWAY_NAME",
                    version: "GATEWAY_VERSION",
                });
                const session = traceImpl(fakeLogger, "ACTION", "MESSAGE");

                // Act
                /**
                 * $FlowExpectedError[cannot-write] Flow is going to complain
                 * because name is readonly.
                 * We will suppress this because we want to test that runtime
                 * code is enforcing this. The idea of this test is to ensure
                 * that folks don't edit out this specific behavior.
                 */
                const underTest = () => (session.action = "NEW_ACTION!");

                // Assert
                expect(underTest).toThrowErrorMatchingInlineSnapshot(
                    `"Cannot set property action of #<Object> which has only a getter"`,
                );
            });
        });

        describe("#addLabel", () => {
            it("should add the label to the span", () => {});
        });

        describe("#end", () => {
            it("should stop the profile timer", () => {
                // Arrange
                const fakeTimer = {
                    done: jest.fn(),
                };
                const fakeLogger: any = {
                    silly: jest.fn(),
                    startTimer: jest.fn().mockReturnValue(fakeTimer),
                };
                jest.spyOn(process, "memoryUsage")
                    .mockReturnValueOnce({
                        someMemoryStuff: 10000,
                    })
                    .mockReturnValueOnce({
                        someMemoryStuff: 50000,
                    });
                jest.spyOn(GetGatewayInfo, "getGatewayInfo").mockReturnValue({
                    name: "GATEWAY_NAME",
                    version: "GATEWAY_VERSION",
                });
                const session = traceImpl(fakeLogger, "ACTION", "MESSAGE");

                // Act
                session.end();

                // Assert
                expect(fakeTimer.done).toHaveBeenCalledWith({
                    message: "TRACED ACTION: MESSAGE",
                    level: "debug",
                    "/memory/delta": {
                        someMemoryStuff: 40000,
                    },
                    "/memory/final": {
                        someMemoryStuff: 50000,
                    },
                });
            });

            it("should use the log level given", () => {
                // Arrange
                const fakeTimer = {
                    done: jest.fn(),
                };
                const fakeLogger: any = {
                    silly: jest.fn(),
                    startTimer: jest.fn().mockReturnValue(fakeTimer),
                };
                jest.spyOn(process, "memoryUsage")
                    .mockReturnValueOnce({
                        someMemoryStuff: 10000,
                    })
                    .mockReturnValueOnce({
                        someMemoryStuff: 50000,
                    });
                jest.spyOn(GetGatewayInfo, "getGatewayInfo").mockReturnValue({
                    name: "GATEWAY_NAME",
                    version: "GATEWAY_VERSION",
                });
                const session = traceImpl(fakeLogger, "ACTION", "MESSAGE");

                // Act
                session.end({level: "silly"});

                // Assert
                expect(fakeTimer.done).toHaveBeenCalledWith({
                    message: "TRACED ACTION: MESSAGE",
                    level: "silly",
                    "/memory/delta": {
                        someMemoryStuff: 40000,
                    },
                    "/memory/final": {
                        someMemoryStuff: 50000,
                    },
                });
            });

            it("should add labels to metadata", () => {
                // Arrange
                const fakeTimer = {
                    done: jest.fn(),
                };
                const fakeLogger: any = {
                    silly: jest.fn(),
                    startTimer: jest.fn().mockReturnValue(fakeTimer),
                };
                jest.spyOn(process, "memoryUsage")
                    .mockReturnValueOnce({
                        someMemoryStuff: 10000,
                    })
                    .mockReturnValueOnce({
                        someMemoryStuff: 50000,
                    });
                jest.spyOn(GetGatewayInfo, "getGatewayInfo").mockReturnValue({
                    name: "GATEWAY_NAME",
                    version: "GATEWAY_VERSION",
                });
                const session = traceImpl(fakeLogger, "ACTION", "MESSAGE");
                session.addLabel("LABEL_A", "label_a");
                session.addLabel("/label/b", "label_b");

                // Act
                session.end();

                // Assert
                expect(fakeTimer.done).toHaveBeenCalledWith({
                    message: "TRACED ACTION: MESSAGE",
                    level: "debug",
                    "/memory/delta": {
                        someMemoryStuff: 40000,
                    },
                    "/memory/final": {
                        someMemoryStuff: 50000,
                    },
                    LABEL_A: "label_a",
                    "/label/b": "label_b",
                });
            });

            it("should use the given metadata info", () => {
                // Arrange
                const fakeTimer = {
                    done: jest.fn(),
                };
                const fakeLogger: any = {
                    silly: jest.fn(),
                    startTimer: jest.fn().mockReturnValue(fakeTimer),
                };
                jest.spyOn(process, "memoryUsage")
                    .mockReturnValueOnce({
                        someMemoryStuff: 50000,
                    })
                    .mockReturnValueOnce({
                        someMemoryStuff: 10000,
                    });
                jest.spyOn(GetGatewayInfo, "getGatewayInfo").mockReturnValue({
                    name: "GATEWAY_NAME",
                    version: "GATEWAY_VERSION",
                });
                const session = traceImpl(fakeLogger, "ACTION", "MESSAGE");

                // Act
                session.end({
                    cached: true,
                    size: 56,
                    someotherrubbish: "blahblahblah",
                });

                // Assert
                expect(fakeTimer.done).toHaveBeenCalledWith({
                    message: "TRACED ACTION: MESSAGE",
                    level: "debug",
                    cached: true,
                    size: 56,
                    someotherrubbish: "blahblahblah",
                    "/memory/delta": {
                        someMemoryStuff: -40000,
                    },
                    "/memory/final": {
                        someMemoryStuff: 10000,
                    },
                });
            });

            it("should override labels with metadata", () => {
                // Arrange
                const fakeTimer = {
                    done: jest.fn(),
                };
                const fakeLogger: any = {
                    silly: jest.fn(),
                    startTimer: jest.fn().mockReturnValue(fakeTimer),
                };
                jest.spyOn(process, "memoryUsage")
                    .mockReturnValueOnce({
                        someMemoryStuff: 10000,
                    })
                    .mockReturnValueOnce({
                        someMemoryStuff: 50000,
                    });
                jest.spyOn(GetGatewayInfo, "getGatewayInfo").mockReturnValue({
                    name: "GATEWAY_NAME",
                    version: "GATEWAY_VERSION",
                });
                const session = traceImpl(fakeLogger, "ACTION", "MESSAGE");
                session.addLabel("cached", false);
                session.addLabel("/label/b", "label_b");

                // Act
                session.end({
                    cached: true,
                    size: 56,
                    someotherrubbish: "blahblahblah",
                });

                // Assert
                expect(fakeTimer.done).toHaveBeenCalledWith({
                    message: "TRACED ACTION: MESSAGE",
                    level: "debug",
                    cached: true,
                    size: 56,
                    someotherrubbish: "blahblahblah",
                    "/memory/delta": {
                        someMemoryStuff: 40000,
                    },
                    "/memory/final": {
                        someMemoryStuff: 50000,
                    },
                    "/label/b": "label_b",
                });
            });

            it("should not allow message to be overridden", () => {
                // Arrange
                const fakeTimer = {
                    done: jest.fn(),
                };
                const fakeLogger: any = {
                    silly: jest.fn(),
                    startTimer: jest.fn().mockReturnValue(fakeTimer),
                };
                jest.spyOn(process, "memoryUsage")
                    .mockReturnValueOnce({
                        someMemoryStuff: 10000,
                    })
                    .mockReturnValueOnce({
                        someMemoryStuff: 50000,
                    });
                jest.spyOn(GetGatewayInfo, "getGatewayInfo").mockReturnValue({
                    name: "GATEWAY_NAME",
                    version: "GATEWAY_VERSION",
                });
                const session = traceImpl(fakeLogger, "ACTION", "MESSAGE");

                // Act
                session.end({
                    message: "I secretly want to change the message",
                });

                // Assert
                expect(fakeTimer.done).toHaveBeenCalledWith({
                    message: "TRACED ACTION: MESSAGE",
                    level: "debug",
                    "/memory/delta": {
                        someMemoryStuff: 40000,
                    },
                    "/memory/final": {
                        someMemoryStuff: 50000,
                    },
                });
            });

            it("should end the trace span", () => {
                // Arrange
                const fakeTraceSpan = {
                    addLabel: jest.fn(),
                    endSpan: jest.fn(),
                };
                const fakeTracer: any = {
                    createChildSpan: jest.fn().mockReturnValue(fakeTraceSpan),
                };
                const fakeTimer = {
                    done: jest.fn(),
                };
                const fakeLogger: any = {
                    silly: jest.fn(),
                    startTimer: jest.fn().mockReturnValue(fakeTimer),
                };
                jest.spyOn(GetGatewayInfo, "getGatewayInfo").mockReturnValue({
                    name: "GATEWAY_NAME",
                    version: "GATEWAY_VERSION",
                });
                const session = traceImpl(
                    fakeLogger,
                    "ACTION",
                    "MESSAGE",
                    fakeTracer,
                );

                // Act
                session.end();

                // Assert
                expect(fakeTraceSpan.endSpan).toHaveBeenCalled();
            });
        });
    });
});
