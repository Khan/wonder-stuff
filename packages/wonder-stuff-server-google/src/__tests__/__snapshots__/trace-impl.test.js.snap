// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#trace should throw if the action is empty/falsy 1`] = `"Must provide an action for the trace session."`;

exports[`#trace should throw if the action is empty/falsy 2`] = `"Must provide an action for the trace session."`;

exports[`#trace should throw if the action is empty/falsy 3`] = `"Must provide an action for the trace session."`;
